sagemathinc/cocalc
# preparations for ubuntu 2004

# Description
this changes a few details in anticipation of switching over to ubuntu 20.04. it's not changing the default software image, that's for later.
* npm: updating babel and antd. I first tried to use a tree selector, but it's not working. still good to update antd.
* typescripting the compute images file, for the standard images. there is now more information per image.
* a distinction between defining the default image (which will change), and a fallback value if it isn't set, which will remain fixed at the image we have now (i.e. `"default"`)
* generalizing/renaming the dialog for selecting the compute image. 
* also introducing grouping of the standard compute images (ubuntu 18.04, "main", ubuntu 20.04, ...) because otherwise it's too repetitve.
* course configuration: it's possible to either change to a particular standard software image or a custom image. there is also more information about what's going on
* changing the above also changes the student projects and the shared project
* removing ugly margins around course panels. in particular, the scrollbars aren't at the border.
* opening/creating a new course sets the compute image, if it isn't already set. that's in `sync.ts`. since I don't know how this works, my code is just a best case. the idea is to set the compute image to the one of the course project, if it isn't already set.
* running files from shared content creates a project using the given standard compute image or custom compute image. it also displays the associated software environment on the shared page.
* only show the compute image selection if the "kucalc ui" is enabled. i.e. there are no compute images in docker or elsewhere. (this applies to creating a new project and configuring the compute image in a course)

# Testing Steps
*  antd related, I checked many dialogs and there are no visible changes
* creating a new course or opening a previously existing one. in both cases the compute image should match the course project's.
* course config: changing the compute image changes student projects and shared project
* creating a new project sets the default, a standard image, or the custom image correctly
* launch/share has 3 cases: nothing is set (use fallback for compute image), standard image is set (e.g. ubuntu or a frozen older one), or a custom image.


# Relevant Issues

### [Checklist](https://github.com/sagemathinc/cocalc/wiki/PR-Checklist):

Front end:
